controllers/pipelinerun_controller.go

```diff
@@ -0,0 +1,203 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// PipelineRunReconciler reconciles PipelineRun objects that may have triggers configured to generate
+// a BuildRun based on the Pipeline state.
+type PipelineRunReconciler struct {
+	client.Client                 // kubernetes client
+	Scheme        *runtime.Scheme // shared scheme
+	Clock                         // local clock instance
+
+	buildInventory *inventory.Inventory // local build triggers database
+}
+
+//+kubebuilder:rbac:groups=shipwright.io,resources=builds,verbs=get;list;watch
+//+kubebuilder:rbac:groups=shipwright.io,resources=buildruns,verbs=create;get;list;update;watch
+//+kubebuilder:rbac:groups=tekton.dev,resources=pipelineruns,verbs=get;list;update;watch
+
+// createBuildRun handles the actual BuildRun creation, uses the informed PipelineRun instance to
+// establish ownership. Only returns the created object name, and error.
+func (r *PipelineRunReconciler) createBuildRun(
+	ctx context.Context,
+	pipelineRun *tknv1beta1.PipelineRun,
+	buildName string,
+) (string, error) {
+	br := v1alpha1.BuildRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    pipelineRun.GetNamespace(),
+			GenerateName: fmt.Sprintf("%s-", pipelineRun.GetName()),
+			Labels: map[string]string{
+				filter.OwnedByPipelineRunLabelKey: pipelineRun.GetName(),
+			},
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: stubs.TektonAPIv1beta1,
+				Kind:       "PipelineRun",
+				Name:       pipelineRun.GetName(),
+				UID:        pipelineRun.GetUID(),
+			}},
+		},
+		Spec: v1alpha1.BuildRunSpec{
+			BuildRef: &v1alpha1.BuildRef{
+				Name: buildName,
+			},
+		},
+	}
+	if err := r.Client.Create(ctx, &br); err != nil {
+		return "", err
+	}
+	return br.GetName(), nil
+}
+
+// triggerBuildsForPipelineRun create the BuildRun instances for the informed objects, and updates
+// the PipelineRun object labels to list the created objects.
+func (r *PipelineRunReconciler) triggerBuildsForPipelineRun(
+	ctx context.Context,
+	logger *logr.Logger,
+	pipelineRun *tknv1beta1.PipelineRun,
+	buildsToBeTriggered []inventory.SearchResult,
+) error {
+	var created []string
+	for _, build := range buildsToBeTriggered {
+		buildRunName, err := r.createBuildRun(ctx, pipelineRun, build.BuildName.Name)
+		if err != nil {
+			return err
+		}
+		created = append(created, buildRunName)
+	}
+	if len(created) == 0 {
+		return fmt.Errorf("no buildruns have been created for %q", pipelineRun.GetNamespacedName())
+	}
+	logger.V(0).Info("BuildRun(s) have been created",
+		"buildruns", created, "pipeline-run-name", pipelineRun.GetNamespacedName())
+
+	// adding a label to the PipelineRun object to identify the BuildRun(s) created for it, and also,
+	// to be able to filter out objects that have already triggered Builds afterwards
+	labels := pipelineRun.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	labels[filter.BuildRunsCreatedKey] = strings.Join(created, ", ")
```

Comment:
Label values have a fairly small length limitation that you will hit here.

Reply:
What's the length limitation? Could we store that information in the PipelineRun status itself?

Reply:
Just 63 characters, https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set. Assuming we do not want to query this data, then we should probably use annotations instead.

Reply:
Please consider the last changes, it's using annotations instead of labels. Thanks!